File: ai/workflows/supportWorkflow.js
import { END, START, StateGraph } from '@langchain/langgraph';
import { BankingState } from '../graph/state/bankingState.js';
import { getWindowToolAction, getWindowToolReply } from '../assistant/responseWrappers.js';
import {
  createExecutedWorkflowResponse,
  createWorkflowResponse
} from '../contracts/assistantResponseContract.js';

const leverageDataNode = async (state, config) => {
  const services = config?.configurable?.services;
  const result = await services.supportService.connectRepresentative({ userId: state.session.userId });
  const workflowResponse = createExecutedWorkflowResponse({
    action: getWindowToolAction('open_video_call_window', result),
    operation: 'open_video_call_window',
    result
  });

  return {
    ...state,
    workflow: { ...state.workflow, activeWorkflow: 'unknown', currentPhase: 'Leverage Data' },
    support: {
      ...state.support,
      ticketId: result?.ticketId || null
    },
    execution: workflowResponse.execution,
    workflowResponse,
    ui: { ...state.ui, action: getWindowToolAction('open_video_call_window', result) }
  };
};

const returnResponseNode = async (state) => {
  const workflowResponse = createWorkflowResponse({
    ...state.workflowResponse,
    message: getWindowToolReply('open_video_call_window', state.session.userLanguage)
  });

  return {
    ...state,
    workflow: { ...state.workflow, currentPhase: 'Return Response with Suggestions' },
    workflowResponse,
    ui: {
      ...state.ui,
      message: workflowResponse.message
    }
  };
};

const supportWorkflowGraph = new StateGraph(BankingState)
  .addNode('leverage_data', leverageDataNode)
  .addNode('return_response_with_suggestions', returnResponseNode)
  .addEdge(START, 'leverage_data')
  .addEdge('leverage_data', 'return_response_with_suggestions')
  .addEdge('return_response_with_suggestions', END)
  .compile();

export const runSupportWorkflow = async ({ state, services }) => (
  supportWorkflowGraph.invoke(state, { configurable: { services } })
);

File: ai/workflows/support_workflow.py
from ai.contracts.assistant_response_contract import create_executed_workflow_response, create_workflow_response
from ai.assistant.response_wrappers import get_window_tool_action, get_window_tool_reply
from observability.langfuse_tracing import duration_ms, get_request_id, now_ms, start_span, trace_log


async def run_support_workflow(*, state: dict, services: dict) -> dict:
    start = now_ms()
    span = start_span(name="support_workflow", metadata={"workflow_name": "support_workflow"})
    support_service = (services or {}).get("supportService")
    user_language = (state.get("session") or {}).get("userLanguage", "en")

    result = await support_service.connect_representative(user_id=(state.get("session") or {}).get("userId"))
    action = get_window_tool_action("open_video_call_window", result)

    workflow_response = create_executed_workflow_response(
        action=action,
        operation="open_video_call_window",
        result=result,
    )

    state = {
        **state,
        "workflow": {**(state.get("workflow") or {}), "activeWorkflow": "unknown", "currentPhase": "Leverage Data"},
        "support": {**(state.get("support") or {}), "ticketId": result.get("ticketId")},
        "execution": workflow_response["execution"],
        "workflowResponse": workflow_response,
        "ui": {**(state.get("ui") or {}), "action": action},
    }

    message = get_window_tool_reply("open_video_call_window", user_language)
    final_response = create_workflow_response(**{**workflow_response, "message": message})

    output = {
        **state,
        "workflow": {**(state.get("workflow") or {}), "currentPhase": "Return Response with Suggestions"},
        "workflowResponse": final_response,
        "ui": {**(state.get("ui") or {}), "message": message},
    }
    ms = duration_ms(start)
    action_type = action.get("type") if isinstance(action, dict) else action
    summary = {
        "workflow_name": "support_workflow",
        "operation": "open_video_call_window",
        "duration_ms": ms,
        "action_type": action_type,
        "status": result.get("status") if isinstance(result, dict) else None,
    }
    span.end(output=summary, metadata=summary)
    trace_log(f"workflow requestId={get_request_id()} name=support_workflow ms={ms:.1f}")
    return output
